### ISSUE TITLE
Support WASM or even Blazor

### ISSUE THREAD

**Issue by FlorianRappl** [smapiot/piral#112]
## Description

WebAssembly should be supported by Parcel out of the box, yet, we may want to improve dev experience here and at least provide a sample if possible.

A potential first target could be client-side-only Blazor. Having .NET developers to supply pilets would be great. One concern here is that if there are multiple pilets are they all shipping a "WASM .NET Runtime"? Can this be shared? How?

## Background

WebAssembly is a standard that is now a W3 recommendation. Blazor is one of the advocated technologies right now from Microsoft, essentially allowing C#/.NET devs to program for the web. It gained traction by the introduction of a WASM enhanced output.

## Useful Links

The following links may or may not be useful:

- https://www.youtube.com/watch?v=OUQH08_y708

## Discussion

Any feedback / suggestions appreciated

**Comment by FlorianRappl** [comment]
WASM in general is not difficult at all.

```ts
import { PiletApi } from "sample-piral";

export function setup(app: PiletApi) {
  import('./fib.wasm').then(({ fib }) => {
    console.log("Result of wasm:", fib(5));
  });
}
```

The problems that I currently see:

- Parcel's WASM loader directly instantiates the WebAssembly, leading to a problem if imports are expected / required.
- Parcel claims it can also "synchronously" work with WebAssembly. This is true for the developer, but not for the module. As such using it with a synchronous import will result in `setup` not found (if done in the main bundle).

All in all I think a new Parcel WASM loader may be helpful here. I'll need to look into details here.

As far as Blazor is concerned: This seems reachable, too - except for the two things mentioned above. However, when I do include it via `fetch` (and manually instantiate it) I get a working blazor pilet!

**Comment by FlorianRappl** [comment]
Here are my thoughts so far:

- Blazor generates quite some files for you, e.g., the WASM loader (JS), the DLL loader (WASM), the manifest (JSON), ...
- Usually, for a standard web app all these things are generated and pre-determined
- Of an empty Blazor client app one may actually include *everything* except the client app's DLL
- This could be the "shared" Blazor runtime
- Writing new Blazor client apps that expose a function just like `setup` would allow having a pilet that consists of a single module only - just exporting `setup`, which just calls something like `piral.fromBlazor('MyBlazor.AppNamespace', fetch('./blazor-client-app.dll'))`
- The `fromBlazor` API would just use the given namespace and the provided DLL (or maybe path to a DLL, even though that may be more complicated to resolve, maybe via `require.resolve('./blazor-client-app.dll')`) to post-load / inject the DLL into the already shared runtime, load a blazorized version of the `setup` function with the `piral` Pilet API object

This all sounds quite complicated, but for the Blazor user it should be as simple as starting / using Blazor with a new kind of Blazor template "Blazor Piral Microfrontend" or just "Blazor Pilet".

So to summarize what's needed for all of this:

- `piral-blazor` to come with the necessary shared runtime and Pilet API extension (same as with `piral-vue` etc.; just convenience)
- A `BlazorPilet` Visual Studio / project template (preferrably installable via `dotnet`)
- A lot of coffee to figure out how to lazy load / post-load DLLs (first sight is not that hard, but I think under production circumstances obstacles will come)
- A good strategy to figure out how to improve / customize debugging / generation of files etc.

Let's see if this can be brought to a PoC first without taking care of the rough edges.

**Comment by FlorianRappl** [comment]
Landed in `develop` and the preview `@next`. We will keep enhancing this.
